c++/src/kj/compat/http.c++

```diff
@@ -5204,65 +5960,77 @@ private:
+          //
+          // Spec: https://tools.ietf.org/html/rfc7231#section-6.3.6
+          connectionHeaders[HttpHeaders::BuiltinIndices::CONTENT_LENGTH] = "0";
+        } else KJ_IF_MAYBE(s, expectedBodySize) {
+          // HACK: We interpret a zero-length expected body length on responses to HEAD requests to
+          //   mean "don't set a Content-Length header at all." This provides a way to omit a body
+          //   header on HEAD responses with non-null-body status codes. This is a hack that *only*
+          //   makes sense for HEAD responses.
+          if (m != HttpMethod::HEAD || *s > 0) {
+            lengthStr = kj::str(*s);
+            connectionHeaders[HttpHeaders::BuiltinIndices::CONTENT_LENGTH] = lengthStr;
+          }
+        } else {
+          connectionHeaders[HttpHeaders::BuiltinIndices::TRANSFER_ENCODING] = "chunked";
+        }
 
-    httpOutput.writeHeaders(headers.serializeResponse(
-        statusCode, statusText, connectionHeadersArray));
+        // For HEAD requests, if the application specified a Content-Length or Transfer-Encoding
+        // header, use that instead of whatever we decided above.
+        kj::ArrayPtr<kj::StringPtr> connectionHeadersArray = connectionHeaders;
+        if (m == HttpMethod::HEAD) {
+          if (headers.get(HttpHeaderId::CONTENT_LENGTH) != nullptr ||
+              headers.get(HttpHeaderId::TRANSFER_ENCODING) != nullptr) {
+            connectionHeadersArray = connectionHeadersArray
+                .slice(0, HttpHeaders::HEAD_RESPONSE_CONNECTION_HEADERS_COUNT);
+          }
+        }
 
-    kj::Own<kj::AsyncOutputStream> bodyStream;
-    if (method == HttpMethod::HEAD) {
-      // Ignore entity-body.
-      httpOutput.finishBody();
-      return heap<HttpDiscardingEntityWriter>();
-    } else if (statusCode == 204 || statusCode == 205 || statusCode == 304) {
-      // No entity-body.
-      httpOutput.finishBody();
-      return heap<HttpNullEntityWriter>();
-    } else KJ_IF_MAYBE(s, expectedBodySize) {
-      return heap<HttpFixedLengthEntityWriter>(httpOutput, *s);
-    } else {
-      return heap<HttpChunkedEntityWriter>(httpOutput);
+        httpOutput.writeHeaders(headers.serializeResponse(
+            statusCode, statusText, connectionHeadersArray));
+
+        kj::Own<kj::AsyncOutputStream> bodyStream;
+        if (m == HttpMethod::HEAD) {
+          // Ignore entity-body.
+          httpOutput.finishBody();
+          return heap<HttpDiscardingEntityWriter>();
+        } else if (statusCode == 204 || statusCode == 205 || statusCode == 304) {
+          // No entity-body.
+          httpOutput.finishBody();
+          return heap<HttpNullEntityWriter>();
+        } else KJ_IF_MAYBE(s, expectedBodySize) {
+          return heap<HttpFixedLengthEntityWriter>(httpOutput, *s);
+        } else {
+          return heap<HttpChunkedEntityWriter>(httpOutput);
+        }
+      }
+      KJ_CASE_ONEOF(m, HttpConnectMethod) {
```

Comment:
I think this shouldn't be possible to reach? We are in the `HttpService::Response::send()` callback here, but on CONNECT requests we don't give `HttpService::Response` to the application in the first place, so it can' be calling this.